ide/ide_slave.ml

```diff
@@ -203,13 +203,23 @@ let export_pre_goals pgs =
     Interface.given_up_goals = pgs.Proof.given_up_goals
   }
 
+let add_diffs oldp newp intf =
+  let open Interface in
+  let (hyps_pp_list, concl_pp) = Proof_diffs.diff_first_goal oldp newp in
+  match intf.fg_goals with
+  | [] -> intf
+  | first_goal :: tl ->
+    { intf with fg_goals = { first_goal with goal_hyp = hyps_pp_list; goal_ccl = concl_pp } :: tl }
+
 let goals () =
+  let oldp = Some (Proof_global.give_me_the_proof ()) in
   let doc = get_doc () in
   set_doc @@ Stm.finish ~doc;
   try
-    let pfts = Proof_global.give_me_the_proof () in
-    Some (export_pre_goals (Proof.map_structured_proof pfts process_goal))
-  with Proof_global.NoCurrentProof -> None
+    let newp = Proof_global.give_me_the_proof () in
+    let intf = export_pre_goals (Proof.map_structured_proof newp process_goal) in
+    Some (add_diffs oldp (Some newp) intf)
+  with Proof_global.NoCurrentProof -> None;;
 
```

Comment:
An expert in the XML protocol should review this; for example, if IDEs do `status` then `goals` this won't work, right?

Reply:
Ah, I don't see the issue, but of course fine to get more review.  The only part I guessed on here is that "set_doc @@ Stm.finish ~doc;" is commiting a proof step and that this is a correct way to get oldp here.  Is that not right?

Reply:
`Stm.finish` may be a noop if the document was already checked.

Reply:
Given the current proof object, is there a way to retrieve the prior version from (I suppose) Stm?  Doing that would solve some other problems (e.g. currently not showing diffs after going backward for lack of the old proof step).  It looked like the proofs used sequential document ids.